test_libs/pyspec/eth2spec/utils/ssz/ssz_partials.py

```diff
@@ -0,0 +1,306 @@
+    o = {starting_index + i: chunk for i, chunk in enumerate(chunks)}
+    o = {**o, **filler(starting_index, len(chunks))}
+    return o
+
+
+@infer_input_type
+def ssz_leaves(obj, typ=None, root=1):
+    if is_list_kind(typ):
+        o = {root * 2 + 1: len(obj).to_bytes(32, 'little')}
+        base = root * 2
+    else:
+        o = {}
+        base = root
+    if is_bottom_layer_kind(typ):
+        data = serialize_basic(obj, typ) if is_basic_type(typ) else pack(obj, read_elem_type(typ))
+        q = {**o, **merkle_tree_of_chunks(chunkify(data), base)}
+        # print(obj, root, typ, base, list(q.keys()))
+        return(q)
+    else:
+        fields = get_typed_values(obj, typ=typ)
+        sub_base = base * next_power_of_two(len(fields))
+        for i, (elem, elem_type) in enumerate(fields):
+            o = {**o, **ssz_leaves(elem, typ=elem_type, root=sub_base + i)}
+        q = {**o, **filler(sub_base, len(fields))}
+        # print(obj, root, typ, base, list(q.keys()))
+        return(q)
+
+
+def fill(objects):
+    objects = {k: v for k, v in objects.items()}
+    keys = sorted(objects.keys())[::-1]
+    pos = 0
+    while pos < len(keys):
+        k = keys[pos]
+        if k in objects and k ^ 1 in objects and k // 2 not in objects:
+            objects[k // 2] = hash(objects[k & - 2] + objects[k | 1])
+            keys.append(k // 2)
+        pos += 1
+    return objects
+
+
+@infer_input_type
+def ssz_full(obj, typ=None):
+    return fill(ssz_leaves(obj, typ=typ))
+
+
+def get_basic_type_size(typ):
+    if is_uint_type(typ):
+        return uint_byte_size(typ)
+    elif is_bool_type(typ):
+        return 1
+    else:
+        raise Exception("Type not basic: {}".format(typ))
+
+
+def get_bottom_layer_element_position(typ, base, length, index):
+    """
+    Returns the generalized index and the byte range of the index'th value
+    in the list with the given base generalized index and given length
+    """
+    assert index < (1 if is_basic_type(typ) else length)
+    elem_typ = typ if is_basic_type(typ) else read_elem_type(typ)
+    elem_size = get_basic_type_size(elem_typ)
+    chunk_index = index * elem_size // 32
+    chunk_count = (1 if is_basic_type(typ) else length) * elem_size // 32
+    generalized_index = base * next_power_of_two(chunk_count) + chunk_index
+    start = elem_size * index % 32
+    return generalized_index, start, start + elem_size
+
+
+@infer_input_type
+def get_generalized_indices(obj, path, typ=None, root=1):
+    if len(path) == 0:
+        return [root] if is_basic_type(typ) else list(ssz_leaves(obj, typ=typ, root=root).keys())
+    if path[0] == '__len__':
+        return [root * 2 + 1] if is_list_type(typ) else []
+    base = root * 2 if is_list_kind(typ) else root
+    if is_bottom_layer_kind(typ):
+        length = 1 if is_basic_type(typ) else len(obj)
+        index, _, _ = get_bottom_layer_element_position(typ, base, length, path[0])
+        return [index]
+    else:
```

Comment:
This can be `elif/else` with jsut a final return, rather than the nesting here.

Reply:
moved partials out of PR